java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java

```diff
@@ -510,14 +510,23 @@ public void testShouldGetDateReturnValidDateWithCalendar() throws Exception {
     Text value = new Text("2021-07-02");
     when(getter.get(0)).thenReturn(value.copyBytes());
 
-    Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("America/Sao_Paulo"));
-    Date result = accessor.getDate(calendar);
+    {
+      Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("America/Sao_Paulo"));
+      Date result = accessor.getDate(calendar);
+      calendar.setTime(result);
 
-    calendar = Calendar.getInstance(TimeZone.getTimeZone("Etc/UTC"));
-    calendar.setTime(result);
+      collector.checkThat(dateTimeFormat.format(calendar.getTime()),
+          equalTo("2021-07-01T21:00:00.000Z"));
```

Comment:
OK. So the assumption is that the underlying value is in UTC, and are converting to the destination time zone?

Derby does something different, though:

```java
    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
    try (final Connection connection = DriverManager.getConnection("jdbc:derby:memory:foo;create=true")) {
      try (final Statement statement = connection.createStatement()) {
        statement.executeUpdate("CREATE TABLE foo (bar DATE)");
        statement.executeUpdate("INSERT INTO foo (bar) VALUES '2021-07-02'");
        try (final ResultSet rs = statement.executeQuery("SELECT * FROM foo")) {
          assertThat(rs.next()).isTrue();
          Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("America/Sao_Paulo"));
          final Date date = rs.getDate(1, calendar);
          System.out.println(date);
          calendar.setTime(date);
          System.out.println(sdf.format(calendar.getTime()));
        }
      }
    }
```

Derby appears to just attach the timezone to the value, instead of adjusting it from an assumed source time zone. The output is:

```
2021-07-02
2021-07-02T12:00:00.000+09:00
```

Reply:
It depends on the whether the timezone field is set in the TimeStampVector.
- If timezone is not set, it assumes UTC and converts from that.
- If timezone is set, it converts from the provided timezone.

If user does not request a different timezone, then the result is good. Otherwise the problem emerges.

Reply:
> If timezone is not set, it assumes UTC and converts from that to target timezone.

This seems a bit off though, since we don't _know_ the timezone in that case specifically. (Not that the current behavior seems any better!) It seems like the right behavior here would be to either treat this as an error, or behave like Derby and treat the user-supplied calendar to mean, "assume the timestamp is a timestamp as viewed in this time zone".

Reply:
Here is the code where it assumes UTC if timezone is missing from the vector: https://github.com/apache/arrow/blob/eaa1a1ea74e2b37020e0133f1460cb251e5f4974/java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessor.java#L180

Yes, I agree with you that we shouldn't blindly assume timezone to be UTC if unset. In Apache Hive, this usually means timestamp without timezone where the values are assumed to be in the default timezone in the JVM (which is the local timezone in the usual case).

Reply:
IMO, any behavior change may need a discussion as it will affect downstream users. This patch only focuses on the bug fix.

Reply:
Fair, though arguably this is replacing buggy behavior with buggy behavior 🙂 

Let's put a caveat in the documentation page for the driver? 

We should consider aligning with what Derby appears to do, too. Either that, or check a few other drivers to see the expected behavior here.

I'm still not sure the conversion here is correct, though. From the docs, java.sql.Date is the millisecond value (in GMT) of midnight of a date in some time zone. 

So if we have the date "2021-07-02" in America/Sao_Paulo, then the GMT timestamp [should be 2021-07-02 03:00](https://www.wolframalpha.com/input?i=2021-07-02+00%3A00+Sao+Paulo+in+GMT). So the new behavior doesn't seem correct and the old behavior is what's expected.

Or if we are assuming that the date is "2021-07-02" in UTC, and converting that to Sao Paulo, that makes no sense - Date isn't supposed to store a time, so we shouldn't treat it as one.

Reply:
The `accessor.getDate(calendar)` above is to implement `ResultSet.getDate(int, Calendar)` and returns a `java.sq.Date`. Given two facts:
- `java.sql.Date`: A milliseconds value represents the number of milliseconds that have passed since January 1, 1970 00:00:00.000 GMT. To conform with the definition of SQL DATE, the millisecond values must be 'normalized' by setting the hours, minutes, seconds, and milliseconds to zero in the particular time zone with which the instance is associated.
- `ResultSet.getDate(int, Calendar)`: Retrieves the value of the designated column in the current row of this ResultSet object as a java.sql.Date object in the Java programming language. This method uses the given calendar to construct an appropriate millisecond value for the date if the underlying database does not store timezone information.

Therefore I agree with you that we should assume the underlying values we are reading do not have timezone information and the behavior would be much easier to understand if we simply attach the user-supplied timezone to the values to return.

The original buggy code that this PR tries to solve actually is not here. The modification here was a result of that fix. Please see my other comments above. Let me double check if there is something wrong.

Reply:
> The original buggy code that this PR tries to solve actually is not here. 

Right, I'm just trying to use this to understand what's happening here.